### ISSUE TITLE
x_defs: allow reading value from a file dependency

### ISSUE THREAD

**Issue by malt3** [bazelbuild/rules_go#3507]
### What version of rules_go are you using?

v0.39.0

### What version of gazelle are you using?

v0.30.0

### What version of Bazel are you using?

v6.1.0

### Does this issue reproduce with the latest releases of all the above?

yes

### What operating system and processor architecture are you using?

linux amd64

### Any other potentially useful information about your toolchain?

n/a

### What did you do?

I have a rule `foo` that creates a file. Let's use a genrule as a placeholder:

```starlark
genrule(
  name = "foo",
  srcs = ["foo.in"],
  outs = ["foo.out"],
  # imagine this producing a meaningful line of text, like a digest
  cmd = "cp $< $@",
)
```

I want to create a go_binary that uses the line of text in a string constant. If possible, I want the string constant to be set at link time.

main.go:
```go
package main

import "fmt"

const replaceme = "this_should_be_replaced_at_linktime"

func main() {
    fmt.Println(replaceme)
}
```

```starlark
go_library(
    name = "main_lib",
    srcs = ["main.go"],
    importpath = "github.com/malt3/project/main",
    visibility = ["//visibility:private"],
    deps = [],
    x_defs = {
        "replaceme": "//foo",
    },
)

go_binary(
    name = "main",
    embed = [":main_lib"],
    visibility = ["//visibility:public"],
)
```

### What did you expect to see?

I would want to have a way to take a correctly formatted[^1] file that is generated during the build and use it to set the value of the x_def. 

### What did you see instead?

AFAICT there is no way to do this currently. I can use stamping if the value is known when running my workspace_status command or hardcode the value but depending on another build step does not seem to work.

### Alternatives I have considered

I can generate a go source file that contains my `const` definition with another build step and use it in the `srcs` attribute. This can be inefficient if other go code depends on the generated source file.

[^1]: A single line of text that can be used as an x_def.

**Comment by fmeum** [comment]
Instead of using `x_defs` for this, couldn't you generate a short `.go` file in a genrule that defines the constant and pass that to the `go_library` in `srcs`?

**Comment by malt3** [comment]
Yes that's what I'm doing now. I'm just afraid that it's more expensive since anything depending on the generated source has to be rebuilt, whereas the x_def can be set at link time.

**Comment by dreamworld** [comment]
Same situation here, I have a generated version file and would like to use x_defs to replace with the file content. Any idea?

**Comment by jscissr** [comment]
I created a PR which implements this feature: https://github.com/bazel-contrib/rules_go/pull/4334

It's not exactly the same as in the original feature request, because the source file is not taken as the entire variable content, but instead it contains variable definitions in the same format as the workspace status. This has the advantage that multiple variables can be in the same file.

Here is an example demonstrating this feature:

BUILD.bazel
```
load("@aspect_bazel_lib//lib:expand_template.bzl", "expand_template")
load("@rules_go//go:def.bzl", "go_binary", "go_library")

expand_template(
    name = "stamp",
    template = ["MY_VARIABLE Hello world!"],
)

go_library(
    name = "main_lib",
    srcs = ["main.go"],
    importpath = "github.com/malt3/project/main",
    stampsrcs = [":stamp"],
    visibility = ["//visibility:private"],
    x_defs = {
        "replaceme": "{MY_VARIABLE}",
    },
)

go_binary(
    name = "main",
    embed = [":main_lib"],
    visibility = ["//visibility:public"],
)

```
main.go
```
package main

import "fmt"

var replaceme = "this_should_be_replaced_at_linktime"

func main() {
    fmt.Println(replaceme)
}
```

**Comment by fmeum** [comment]
What I missed in my original reply is to recommend `//go:embed` as an alternative, which allows you to embed the contents of a file. It wouldn't directly give you the values of multiple variables, but you could parse them out of the file in regular, user-controlled Go.

@jscissr Have you considered that approach? If so, what makes you prefer `stampsrcs`?
